src/Symfony/Component/Translation/Util/ArrayConverter.php

```diff
@@ -96,4 +97,45 @@ private static function cancelExpand(array &$tree, string $prefix, array $node)
             }
         }
     }
+
+    private static function getKeyParts(string $key)
```

Comment:
I think we can simplify this method with something like:

```
function getKeyParts(string $key) {
    $parts = explode('.', $key);
    $nb = count($parts);

    if (str_starts_with($key, '.')) {
        $parts[0] = '.';
    }
    if (str_ends_with($key, '.')) {
        $parts[$nb-1] = '.';
    }
    return $parts;
}
```

Reply:
Thanks @welcoMattic. I love short code...

Nevertheless the shared code has 2 small disadvantages:

1. It is using functions available from php 8.0^ (str_starts_with and, str_ends_with), this code is intended to be part of the version 5.4 which still supports older versions of php.
2. Even if this covers the main goal (which is prevent empty keys with dots at the begining and, in the end) this is not covering some edge cases:

```
'abc 1.abc 2' => 'value',
'bcd 1.bcd 2.' => 'value',
'.cde 1.cde 2.' => 'value',
'.def 1.def 2' => 'value',
```

Please see [the test](https://github.com/symfony/symfony/pull/52005/files#diff-972cebacde3182f368ea028839c181153a8ba996565dc88c4854e65b8a6bd343R90) for this section for more info.

Reply:
for php version, you can leverage [symfony/polyfill-php](https://packagist.org/search/?query=symfony%2Fpolyfill-php) (find the one v that introduced the relevant function)

Reply:
The polyfill that introduces `str_starts_with` and `str_ends_with` is [symfony/polyfill-php80](https://packagist.org/packages/symfony/polyfill-php80), and it is already required in the Translation component so using those functions is ok imo

Reply:
ho indeed cool if already included

Reply:
@javleds It seems the edge cases you mention are covered by simplier code here:
https://3v4l.org/2ic9k
```php
function getKeyParts(string $key) {
    $parts = explode('.', $key);
    $nb = count($parts);

    if (str_starts_with($key, '.')) {
        $parts[0] = '.';
    }
    if (str_ends_with($key, '.')) {
        $parts[$nb-1] = '.';
    }
    return $parts;
}


var_dump(getKeyParts('abc 1.abc 2'));
var_dump(getKeyParts('bcd 1.bcd 2.'));
var_dump(getKeyParts('.cde 1.cde 2.'));
var_dump(getKeyParts('.def 1.def 2'));
```

Reply:
@javleds It seems the edge cases you mention are covered by simplier code here:
https://3v4l.org/2ic9k
```php
function getKeyParts(string $key) {
    $parts = explode('.', $key);
    $nb = count($parts);

    if (str_starts_with($key, '.')) {
        $parts[0] = '.';
    }
    if (str_ends_with($key, '.')) {
        $parts[$nb-1] = '.';
    }
    return $parts;
}


var_dump(getKeyParts('abc 1.abc 2'));
var_dump(getKeyParts('bcd 1.bcd 2.'));
var_dump(getKeyParts('.cde 1.cde 2.'));
var_dump(getKeyParts('.def 1.def 2'));
```

Reply:
> Thanks for the advise of php polyfill, I will keep in mind in the case we need to perform additional udates. 🙏 

Getting back to the code discussion:

The real problem here is to avoid trimming the periods at the beginning/end of the parts. Since the `getElementByPath` function is itended to process each part as a valid string. Having empty positions or a position wihtn a single dot won't give the expected result.

For example:
- `['foo', 'bar']` will give the expected result.
- `['foo', '', 'bar']` won't give the expected result due the empty position.
- `['foo', 'bar', '.']` won't give the expected result due the period at position 2.
- `['.foo', 'bar.']` will give the expected result.

The simpler solution is to keep the period concatenated to the key part, and prevent the period occupate a single possition in the array. Otherwise, we will have to add extra processing inside the `getElementByPath` function.

Having this in mind, the code proposed above has small variations on the expected result.

Let me give the examples:

| input     | expected         | output with current fix | output with simpler code |
|-----------|------------------|-------------------------|--------------------------|
| abc.abc   | ['abc', 'abc']   | ['abc', 'abc']          | ['abc', 'abc']           |
| bcd.bcd.  | ['bcd', 'bcd.']  | ['bcd', 'bcd.']         | ['bcd', 'bcd', '.']      |
| .cde.cde. | ['.cde', 'cde.'] | ['.cde', 'cde.']        | ['.', 'cde', 'cde', '.'] |
| .def.def  | ['.def', 'def']  | ['.def', 'def']         | ['.','def', 'def']       |


If the proposed solution is hard to read, maybe, we can create named variables for every `if` statement.